File: homeworks/module2/asian-option.py
import numpy

def lin_interp(u, ax, n1, n2, da, nsize):
    """ 
	function implements linear interpolation:
	fx = f1 + (f2-f1)/(a2-a1)*(ax-a1)
    
    parameters
    ----------
    u : FD solution at the time-step t(i) (i.e. just after)
	ax : grid point just before a[x]
	n1 : grid point just before a[x]
	n2 : grid point just after a[x]
    da : increment in A dimention
	nsize: size of the spatial grid in S dimention
    
    returns
    -------
    vn : interpolated value at the time-step t(i)^- (i.e. just before)
    """
    for i in range(1, nsize):
		ux[i, k] = u[i, n1] + (u[i, n2] - u[i, n1]) / (n2 - n1) * (ax / da - n1)

	return ux
	
def solve_lin_system(Amatrix, b):
    """ 
	function implements solution of linear system Ax  = b:
    
    parameters
    ----------
	Amatrix : 
	b : 
  
    returns
    -------
    vn : interpolated value at the time-step t(i)^- (i.e. just before)
    """
	y = Amatrix*x + b 
	
	return x	
	
# initial parameters
T = 30 / 365
r = 0.05
sigma = 0.40
K = 100.12
S0 = 100.0

# set-up containers
dt = 0.01  # T - time
da = 0.02  # A - average price
ds = 0.03  # S - spot price

# pricing region (Smin;Smax)
Smax = K + 3*sigma*S0
Smin = max(K - 3*sigma*S0,0.0)

nt = int(T / dt) + 1  # T - time dimension
na = int((Smax - Smin) / da) + 1  # A = (Smin;Smax) - average price dimension
ns = int((Smax - Smin) / ds) + 1  # S = (Smin;Smax) - spot price dimension

v = numpy.zeros((ns, na))
Amatrix = numpy.zeros((ns-1, ns-1))
b = numpy.zeros((ns-1))

# set-up I.C. at time t = T : v(A,S,T) = max(A-K,0.0)
for k in range(1, na):
    for i in range(1, ns):
        v[i, k] = max(na * k - K, 0.0)

		# we are stepping backwards:
for n in range(nt, 1, -1):
	
	# # set BC-1: if Smin << K, then A << K, so v(A,Smin,t) = max(A-K,0.0) = 0.0
    # for k in range(1, na):
		# v[0, k] = 0.0 
	
	# # BC-2: if Smax >> K, d^2V/dS^2 == 0, so dV/dt + rSdV/dS + 0.5*sigma^2*0.0 = rV	
	# for k in range(1, na):
		# v[ns,k] = 0.0 
	
	v_old = v.copy()
    for k in range(1, na):
        # a[x] = a[k] + (s[j]-a[k])/t(n)
        ax = da * k + (ds * j - da * k) / (dt * n)

        # determine grid point just before and just after a[x]
        nx = ax / da
        n1 = int(nx)
        n2 = n1 + 1

        # implement jump condition by using linear interpolation:
        # fx = f1 + (f2-f1)/(a2-a1)*(ax-a1)
		vn = lin_interp(v_old, ax, n1, n2, da, ns)

    # implement Black-Scholes equation backward time stepping
	# prepare matrix A and vector b

	### i == 0
	# set BC-1: if Smin << K, then A << K, so v(A,Smin,t) = max(A-K,0.0) = 0.0
	v[0] = 0.0 
        
	### 0 < i < ns
	# implement Black-Scholes discretization:
	# v(tn+1,i+1) + (1 -alpha + 2*gamma)/(beta + gamma)*v(tn+1,i) + v(tn+1,i-1) = b
		
	# where		
	# alpha = 0.5*r*dt
	# beta = 0.5*r*S(i)*dt/dx
	# gamma = 0.5*sigma^2*S(i)^2*dt/(dx)^2
	# b = v(tn,i+1)*(-1) + v(tn,i)*(1 + alpha + 2*gamma)/(beta + gamma) + v(tn,i-1)*(beta - gamma)/(beta + gamma) 
	for i in range(1, ns-1):
		Amatrix[i-1,i] = 1.0
		Amatrix[i,i] = (1 -alpha + 2*gamma)/(beta + gamma)
		Amatrix[i+1,i] = 1.0
		b[i] = v[i+1]*(-1) + v[i]*(1 + alpha + 2*gamma)/(beta + gamma) + v[i-1]*(beta - gamma)/(beta + gamma)
		
	### i == ns
	# BC-2: if Smax >> K, d^2V/dS^2 == 0, so dV/dt + rSdV/dS + 0.5*sigma^2*0.0 = rV			
	# v(tn+1,i) + beta*v(tn+1,i+1) - beta*v(tn+1,i) = c
		
	# where
	# beta = 0.5*r*S(i)*dt/dx
	# c = r*v(tn,i)*dt - beta*v(tn,i+1) +beta*v(tn,i)
	Amatrix[ns-1,ns] = beta
	Amatrix[ns,ns] = -beta
	b[ns] = c
		
	v_out = solve_lin_system(Amatrix, b)
	
print 'run everything'

File: homeworks/module2/asian_option.m
% initial parameters
T = 365.0 / 365.0;
r = 0.10;
sigma = 0.40;
K = 100.0;
S0 = 105.0;
Nsamples = 365;

% set-up containers
nt = 30;
ns = 10;
ng = 14;

% pricing region (Smin;Smax)
Smax = S0 + 3 * sigma * T * S0;
Smin = max(S0 - 3 * sigma * T * S0, 0.0);

% running sum region (Gmin;Gmax)
Gmax = K + Smax*Nsamples;
Gmin = Smin;

dt = T / (nt - 1);
ds = (Smax - Smin) / (ns - 1);
dg = (Gmax - Gmin) / (ng - 1);

vv = zeros(ng,ns);

% n = nt

% set-up I.C. at time t = T : v(S,g,T) = max(g/Nsample-K,0.0)
for j = 1:ng
	for i = 1:ns
		vv(j,i) = max(dg * j/ Nsamples - K, 0);
	end;
end;

% we are stepping backwards:
for n = nt-1:-1:1
	for j = 1:ng
		v_old = vv(j,:);

		% prepare matrix A and vector b
		Amatrix = zeros(ns, ns);
		b = zeros(ns,1);

		% idx == 0 -> Smin
		% BC-1: if Smin << K, d^2V/dS^2 == 0, so dV/dt + rSdV/dS + 0.5*sigma^2*0.0 = rV
		% v(tn-1,i) = v(tn,i)*(1 - alpha0 - beta0) + v(tn-1,i+1)*beta0

		% where
		% alpha0 = r*dt
		% beta0 = r*S(i)*dt/dx
		% c0 = v(tn,i)*(1 - alpha0 - beta0) + v(tn-1,i+1)*beta0
		idx = 1;
		alpha0 = r * dt;
		beta0 = r * (idx * ds) * dt / ds;

		c0 = v_old(idx) * (1.0 - alpha0 - beta0) + v_old(idx + 1) * beta0;

		Amatrix(idx, idx) = 1.0;
		b(idx) = c0;
		
		% idx == ns -> Smax
		% BC-2: if Smax >> K, d^2V/dS^2 == 0, so dV/dt + rSdV/dS + 0.5*sigma^2*0.0 = rV
		% v(tn-1,i) = v(tn,i)*(1 - alpha1 + beta1) + v(tn-1,i-1)*beta1
		
		% where
		% alpha1 = r*dt
		% beta1 = r*S(i)*dt/dx
		% c1 = v(tn,i)*(1 - alpha1 + beta1) + v(tn-1,i-1)*beta1
		
		idx = ns;
		alpha1 = r * dt;
		beta1 = r * (idx * ds) * dt / ds;

		c1 = v_old(idx) * (1.0 - alpha1 + beta1) - v_old(idx - 1) * beta1;

		Amatrix(idx, idx) = 1.0;
		b(idx) = c1;

		% %% 0 < idx < ns-1
		% implement Black-Scholes discretization:
		% v(tn+1,i+1) + (1 -alpha + 2*gamma)/(beta + gamma)*v(tn+1,i) + v(tn+1,i-1) = b
		% where
		% alpha2 = 1/2*r*dt
		% beta2 = 1/4*r*S(i)*dt/dx
		% gamma2 = 1/8*sigma^2*S(i)^2*dt/(dx)^2
		% c2 = v_old(i+1)*(-1) + v_old(i)*(alpha2 + 2*gamma2 - 1.0)/(beta2 + gamma2) + v_old(i-1)*(gamma2 - beta2)/(beta2 + gamma2)
		for i = 1+1:ns-1
			alpha2 = 1 / 2. * r * dt;
			beta2 = 1 / 4. * r * (ds * i) * dt / ds;
			gamma2 = 1 / 8. * sigma^2 * (ds * i)^2 * dt / ds^2;
			c2 = v_old(i + 1) * (-1) + v_old(i) * (alpha2 + 2 * gamma2 - 1.0) / (beta2 + gamma2) + v_old(i - 1) * (beta2 - gamma2) / (beta2 + gamma2);

			Amatrix(i, i - 1) = 1.0;
			Amatrix(i, i) = -(1 + alpha2 + 2 * gamma2) / (beta2 + gamma2);
			Amatrix(i,i + 1) = (gamma2 - beta2) / (beta2 + gamma2);
			b(i) = c2;
		end;
		vv(j,:) = Amatrix\b;
	end;

	% implement jump condition V(S,g,ti^-) = V(S,g + S,ti)
	vv_copy = vv;

	for j = 1:ng
		for i = 1:ns
			jj = floor(j + i*ds/dg);
			% if for cases when g+S > Gmax set value of v same as for Gmax 
			if(jj+1 > ng)		
				vv(j,i) = vv_copy(ng,i);
			else
				vv(j,i) = vv_copy(jj,i) + (vv_copy(jj+1,i) - vv_copy(jj,i))*(j +i*ds/dg -jj) ;
			end;	
		end;
	end;	
end;

% validation step
csvwrite('vv.csv',vv)
mesh(vv)

Gx = 0.0;
Sx = 105.0;

% x = Smin + (1:ns)*ds;
% y = vv(1,:);

% option_price = spline(x,y,Sx);

% plot(x,y);
% hold on
% plot(Sx,option_price,'o');
% hold off

% option_price = get_option_price(vv,[Gx,Sx],[dg,ds],[Gmin,Smin])